### PR TITLE
[clang][Modules] Remove unnecessary includes of `Module.h`

### PR DESCRIPTION
**Pull request by davidstone**

### CODE REVIEW

File: clang/include/clang/APINotes/APINotesManager.h [hunk 1]
```diff
@@ -159,7 +159,8 @@ class APINotesManager {
   ArrayRef<APINotesReader *> getCurrentModuleReaders() const {
     bool HasPublic = CurrentModuleReaders[ReaderKind::Public];
     bool HasPrivate = CurrentModuleReaders[ReaderKind::Private];
-    assert((!HasPrivate || HasPublic) && "private module requires public module");
+    assert((!HasPrivate || HasPublic) &&
+           "private module requires public module");
```
Comment (ChuanqiXu9) [inline clang/include/clang/APINotes/APINotesManager.h:163]: Please don't change untouched codes.

File: clang-tools-extra/clangd/unittests/ReplayPeambleTests.cpp [hunk 2]
```diff
@@ -42,7 +41,11 @@
 #include <memory>
 #include <vector>
 
-namespace clang::clangd {
+namespace clang {
+
+class Module;
```
Comment (kadircet) [inline clang-tools-extra/clangd/unittests/ReplayPeambleTests.cpp:46]: this is not an "unnecessary" include, clangd requires a declaration of `clang::Module` in this file (even if incomplete) and we try not to have forward declarations.

what's the motivation behind this change exactly?

**Review by kadircet — CHANGES_REQUESTED** [review]

File: clang/include/clang/APINotes/APINotesManager.h (hunk 1 above)
Comment (davidstone) [reply clang/include/clang/APINotes/APINotesManager.h:163]: clang-format did that when I touched the file. Does the formatting requirement apply to a narrower set of changes than the entire file?

File: clang-tools-extra/clangd/unittests/ReplayPeambleTests.cpp (hunk 2 above)
Comment (davidstone) [reply clang-tools-extra/clangd/unittests/ReplayPeambleTests.cpp:46]: Is that a clangd preference or llvm-wide? I see a lot of forward declarations in other parts of the code so I want to understand when I should and shouldn't use them in this project.

The motivation is that I'm making a bunch of changes to Module.h, and reducing the number of files that include it speeds up my local development (and full builds, too) by requiring less rebuilding.

Comment (davidstone) [reply clang-tools-extra/clangd/unittests/ReplayPeambleTests.cpp:46]: Specifically I'm following the guidance at: https://llvm.org/docs/CodingStandards.html#include-as-little-as-possible

File: clang/include/clang/APINotes/APINotesManager.h (hunk 1 above)
Comment (ChuanqiXu9) [reply clang/include/clang/APINotes/APINotesManager.h:163]: Yes, we prefer to format the changed line only. Otherwise the backporting may be problematic. And git blaming will be harder.

One possible way maybe:

> git diff -U0 --no-color --relative HEAD^ | clang/tools/clang-format/clang-format-diff.py -p1 -i

File: clang-tools-extra/clangd/unittests/ReplayPeambleTests.cpp (hunk 2 above)
Comment (kadircet) [reply clang-tools-extra/clangd/unittests/ReplayPeambleTests.cpp:46]: yes that's mostly something we try in clangd, but fair point, feel free to ignore this one.

i was mostly pushing because this is just a leaf hence having that dependency won't affect compile times drastically, and we'll probably keep pulling it through other headers anyway. (e.g. even after this change, we still have `clangd/SourceCode.h -> clang/Lex/HeaderSearch.h -> clang/Lex/DirectoryLookup.h -> clang/Lex/ModuleMap.h -> clang/Basic/Module.h, hence this isn't really changing much in practice, while hiding the dependency)

**Review by ldionne — APPROVED** [review]
libcxx/ changes LGTM.

File: clang/include/clang/APINotes/APINotesManager.h (hunk 1 above)
Comment (davidstone) [reply clang/include/clang/APINotes/APINotesManager.h]: OK, I set my VS Code to just format changed lines, looks like I have it working now.

**Comment by davidstone** [comment]
Is there any additional work needed on this before it can be merged?

**Review by ChuanqiXu9 — APPROVED** [review]
LGTM

**Comment by ChuanqiXu9** [comment]
> Is there any additional work needed on this before it can be merged?

No, it looks good. I'll merge this.
